---
title: "predicting_breast_cancer_survival"
author: "Serdar Korur"
date: "12/30/2019"
output: html_document
---

```{r, message=FALSE, warning=FALSE}

library(tidyverse)
library(cgdsr)
# library(TCGA-Assembler)
```

```{r. guide}
# Create CGDS object by CGDS function
mycgds = CGDS("https://www.cbioportal.org/")
test(mycgds)

# Get list of cancer studies at server
getCancerStudies(mycgds)

# Get available case lists (collection of samples) for a given cancer study
mycancerstudy = getCancerStudies(mycgds)[2,1]
mycaselist = getCaseLists(mycgds,mycancerstudy)[1,1]

# Get available genetic profiles
mygeneticprofile = getGeneticProfiles(mycgds,mycancerstudy)[4,1]

# Get data slices for a specified list of genes, genetic profile and case list
getProfileData(mycgds,c('BRCA1','BRCA2'),mygeneticprofile,mycaselist)

# Get clinical data for the case list
myclinicaldata = getClinicalData(mycgds,mycaselist)

# documentation
help('cgdsr')
help('CGDS')

```


# Looking for breast cancer data

```{r, search breast cancer studies, Whole-exome sequencing of 237 metastatic breast cancer samples}

# List all studies
cancerlist <- getCancerStudies(mycgds)

# To find studies where breast is mentioned in the description
breast <- str_which(cancerlist$name, "[Bb]reast")
cancerlist[breast,]

# Get Clinical data for the breast cancer study no 252
# The Metastatic Breast Cancer Project (Provisional, October 2018
mycancerstudy = getCancerStudies(mycgds)[252,1]
mycaselist = getCaseLists(mycgds,mycancerstudy)[1,1]
myclinicaldata = getClinicalData(mycgds,mycaselist)
dim(myclinicaldata)
# view(myclinicaldata)

# subset data

myc <- myclinicaldata[,c("MUTATION_COUNT", "FRACTION_GENOME_ALTERED" )]

# Visualize
myc %>% ggplot(aes(x=FRACTION_GENOME_ALTERED, y=MUTATION_COUNT)) + geom_point() + geom_smooth()

#fit a linear model
lmout <- lm(formula = FRACTION_GENOME_ALTERED ~ MUTATION_COUNT, data = myc)
summary(lmout)
anova(lmout) # There is a significant relation between mutation count and fraction of genome altered
```


```{r, check another breast cancer dataset}
# https://www.cbioportal.org/datasets

# List all studies
cancerlist <- getCancerStudies(mycgds)

# To find studies where breast is mentioned in the description
breast <- str_which(cancerlist$name, "[Bb]reast")
cancerlist[breast,]

# Get Clinical data for the breast cancer study no 252
# The Metastatic Breast Cancer Project (Provisional, October 2018
mycancerstudy = getCancerStudies(mycgds)[46,1]
mycaselist = getCaseLists(mycgds,mycancerstudy)[1,1]
myclinicaldata = getClinicalData(mycgds,mycaselist)
dim(myclinicaldata)
# view(myclinicaldata)



# subset data
myc <- myclinicaldata[,c("MUTATION_COUNT", "FRACTION_GENOME_ALTERED" )]
myc <- myc %>% drop_na(MUTATION_COUNT)


# Visualize

myc %>% ggplot(aes(x=FRACTION_GENOME_ALTERED, y=MUTATION_COUNT)) + geom_point() + geom_smooth()

# fit a linear model

lmout <- lm(formula = FRACTION_GENOME_ALTERED ~ MUTATION_COUNT, data = myc)
lmout
anova(lmout) # There is a significant relation between mutation count and fraction of genome altered
```


# Average mutation count in different cancers

```{r}

```


```{r, Engineer multiple datasets into a target dataframe}
# for loop
myc <- list()
myclinicaldata <- list()
index <- setdiff(1:273, c(48,137,140,163))
for (i in index){
mycancerstudy <- getCancerStudies(mycgds)[i,1]
mycaselist <- getCaseLists(mycgds,mycancerstudy)[1,1]
myclinicaldata <- getClinicalData(mycgds,mycaselist)
if("FRACTION_GENOME_ALTERED" %in% names(myclinicaldata)){
   if("MUTATION_COUNT" %in% names(myclinicaldata)){
  myc[[i]] <- myclinicaldata[,c("MUTATION_COUNT", "FRACTION_GENOME_ALTERED" )]
   }
}
}
vec <- lapply(myc,is.null)
datafrac <- which(vec==FALSE)


# Combine whole dataset
myc2 <- myc[datafrac]
for (i in 1:44) {
  myc2[[i]] <- myc2[[i]] %>% mutate(ID = datafrac[i], study = getCancerStudies(mycgds)[datafrac[i],2])
}

mycall <- do.call(rbind, myc2)
mycall <- mycall %>% drop_na()
dim(mycall)
# Count mean number of mutations in whole data
mean(mycall$MUTATION_COUNT)
mean(mycall$FRACTION_GENOME_ALTERED)
# Count mean number of mutations in each study
mycall %>% group_by(ID, study) %>% summarise(n = mean(MUTATION_COUNT)) %>% arrange(-n)
mycall %>% group_by(ID, study) %>% summarise(n = mean(FRACTION_GENOME_ALTERED)) %>% arrange(-n)
# Create a data frame for each study with mutation cou8nt and fraction genome altere
sum_frac_mut <- mycall %>% group_by(ID, study) %>% summarise(nfrac = mean(FRACTION_GENOME_ALTERED), nmut = mean(MUTATION_COUNT)) 
# Plot
sum_frac_mut %>% ggplot(aes(nfrac, log(nmut), color = study)) + geom_point() + geom_label_repel(aes(label=ID, size =0.1)) + theme(legend.position = "none") # all studies
# lung studies
sum_frac_mut %>% ggplot(aes(nfrac, log(nmut), color = study)) + geom_point() + geom_label_repel(aes(label=ID, color = ifelse(grepl('Lung', study), "red", "blue") + theme(legend.position = "none")
                                                                                                    
```


```{r}
 geom_label_repel(aes(label=ID, color = ifelse(grepl('Lung', study), "red", "blue")))
```


```{r, modelling multiple datasets}
#mycall$FRACTION_GENOME_ALTERED <- scale(mycall$FRACTION_GENOME_ALTERED)

## apply a linear model
lmfull <- lm(formula = FRACTION_GENOME_ALTERED ~ MUTATION_COUNT, data = mycall)
## apply a linear mixed model
lmerfull <- lmer(FRACTION_GENOME_ALTERED ~ MUTATION_COUNT + (1 | ID), data = mycall)


# Apply linear model to all datasets which contain both variables "MUTATION_COUNT", "FRACTION_GENOME_ALTERED"
lmout <- list()
for (i in datafrac) {
mycancerstudy <- getCancerStudies(mycgds)[i,1]
mycaselist <- getCaseLists(mycgds,mycancerstudy)[1,1]
myclinicaldata <- getClinicalData(mycgds,mycaselist)
myc <- myclinicaldata[,c("MUTATION_COUNT", "FRACTION_GENOME_ALTERED" )]
myc <- myc %>% drop_na()

lmout[[i]] <- lm(formula = FRACTION_GENOME_ALTERED ~ MUTATION_COUNT, data = myc)

}

a <- list()
for (i in datafrac) {
  a[[i]] <- summary(lmout[[i]])$adj.r.squared * 100
}
```

```{r, test}

# for loop
myc <- list()
myclinicaldata <- list()
for (i in 1:10){
mycancerstudy <- getCancerStudies(mycgds)[i,1]
mycaselist <- getCaseLists(mycgds,mycancerstudy)[1,1]
myclinicaldata <- getClinicalData(mycgds,mycaselist)
if("FRACTION_GENOME_ALTERED" %in% names(myclinicaldata)){
   if("MUTATION_COUNT" %in% names(myclinicaldata)){
  myc[[i]] <- myclinicaldata[,c("MUTATION_COUNT", "FRACTION_GENOME_ALTERED" )]
   }
}
}

vec <- lapply(myc,is.null)
datafrac <- which(vec==FALSE)
myc <- list()

#second for loop
lmout <- list()
for (i in datafrac) {
mycancerstudy <- getCancerStudies(mycgds)[i,1]
mycaselist <- getCaseLists(mycgds,mycancerstudy)[1,1]
myclinicaldata <- getClinicalData(mycgds,mycaselist)
myc <- myclinicaldata[,c("MUTATION_COUNT", "FRACTION_GENOME_ALTERED" )]
myc <- myc %>% drop_na(MUTATION_COUNT)

lmout[[i]] <- lm(formula = FRACTION_GENOME_ALTERED ~ MUTATION_COUNT, data = myc)

}

a <- list()
for (i in datafrac) {
  a[[i]] <- summary(lmout[[i]])$adj.r.squared * 100
}
```

compare models with anova()

lmout <- lm(MUTATION_COUNT ~ FRACTION_GENOME_ALTERED, data = mycall)
lmer1 <- lmer(MUTATION_COUNT ~ FRACTION_GENOME_ALTERED + (1|ID), data = mycall)
lmer2 <- lmer(MUTATION_COUNT ~ (FRACTION_GENOME_ALTERED|ID),data = mycall)
lmer3 <- lmer(MUTATION_COUNT ~ FRACTION_GENOME_ALTERED + (FRACTION_GENOME_ALTERED|ID),data = mycall)
lmer4 <- lmer(MUTATION_COUNT ~ FRACTION_GENOME_ALTERED + (1 + FRACTION_GENOME_ALTERED|ID),data = mycall)

use AIC(lmout,lmer1,lmer2,lmer3,lmer4)

Results: Cancer study ID explained 30% of variance in mutation count, 5% of variance explained by the FRACTION_GENOME_ALTERED
```{r, lmer model for predicting mutation_count with fraction of genome altered}
library(lme4)

# models
```

# Resources

**Bioinformatic Tools**
https://icahn.mssm.edu/research/bioinformatics/resources
https://www.coursera.org/lecture/network-biology/classifying-patients-tumors-from-tcga-Eeej3?authMode=signup&redirectTo=%2Flearn%2Fnetwork-biology%3Faction%3Denroll

**machine learning algorithm for predicting pfs**

https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-019-2929-8

**use of GAMs to categorise contnious variables in lung disease prediction**

https://bmcmedresmethodol.biomedcentral.com/articles/10.1186/1471-2288-13-83

**Useful packages**
https://www.nature.com/articles/s41597-019-0207-2#Fig1