---
title: "predicting_breast_cancer_survival"
author: "Serdar Korur"
date: "12/30/2019"
output: html_document
---

```{r, message=FALSE, warning=FALSE}

library(tidyverse)
library(cgdsr)
# library(TCGA-Assembler)
```

```{r. guide}
# Create CGDS object
mycgds = CGDS("https://www.cbioportal.org/")

test(mycgds)

# Get list of cancer studies at server
getCancerStudies(mycgds)

# Get available case lists (collection of samples) for a given cancer study
mycancerstudy = getCancerStudies(mycgds)[2,1]
mycaselist = getCaseLists(mycgds,mycancerstudy)[1,1]

# Get available genetic profiles
mygeneticprofile = getGeneticProfiles(mycgds,mycancerstudy)[4,1]

# Get data slices for a specified list of genes, genetic profile and case list
getProfileData(mycgds,c('BRCA1','BRCA2'),mygeneticprofile,mycaselist)

# Get clinical data for the case list
myclinicaldata = getClinicalData(mycgds,mycaselist)

# documentation
help('cgdsr')
help('CGDS')

```


# Looking for breast cancer data

```{r, search breast cancer studies, Whole-exome sequencing of 237 metastatic breast cancer samples.}

# List all studies
cancerlist <- getCancerStudies(mycgds)

# To find studies where breast is mentioned in the description
aaa <- str_which(cancerlist$description, "breast")
cancerlist[aaa]

# Get Clinical data for the breast cancer study no 252
# The Metastatic Breast Cancer Project (Provisional, October 2018
mycancerstudy = getCancerStudies(mycgds)[252,1]
mycaselist = getCaseLists(mycgds,mycancerstudy)[1,1]
myclinicaldata = getClinicalData(mycgds,mycaselist)
view(myclinicaldata)

# subset data

myc <- myclinicaldata[,c("MUTATION_COUNT", "FRACTION_GENOME_ALTERED" )]

# Visualize
myc %>% ggplot(aes(x=FRACTION_GENOME_ALTERED, y=MUTATION_COUNT)) + geom_point() + geom_smooth()

#fit a linear model
lmout <- llm(formula = FRACTION_GENOME_ALTERED ~ MUTATION_COUNT, data = myc)
lmout
anova(lmout) # There is a significant relation between mutation count and fraction of genome altered
```

# Resources

**machine learning algorithm for predicting pfs**

https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-019-2929-8

**use of GAMs to categorise contnious variables in lung disease prediction**

https://bmcmedresmethodol.biomedcentral.com/articles/10.1186/1471-2288-13-83

**Useful packages**
https://www.nature.com/articles/s41597-019-0207-2#Fig1